File: smartlist/src/main/java/com/smartlist/module/grading/GradingController.java
package com.smartlist.module.grading;

import com.smartlist.module.grading.dto.request.QuestionCommentReqDTO;
import com.smartlist.module.grading.dto.request.QuestionScoreReqDTO;
import com.smartlist.module.grading.dto.response.AssessmentQuestionScoreRespDTO;
import com.smartlist.module.grading.dto.response.QuestionCommentRespDTO;
import com.smartlist.enums.AssessmentDecision;
import com.smartlist.utils.AppResponseCodes;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.ArraySchema;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.Positive;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.annotation.Secured;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@Tag(name = "Grading", description = "Grading API")
@RequestMapping(value ="/grading/", produces = { "application/json" })
@ResponseStatus(HttpStatus.OK)
@Validated
@Secured({"SUPERADMIN", "ADMIN", "MEMBER"})
public class GradingController {

    private final GradingService gradingService;

    GradingController(final GradingService gradingService){
        this.gradingService = gradingService;
    }

    @Operation(summary = "Get candidate question scores for assessment")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", array = @ArraySchema(schema = @Schema(implementation = AssessmentQuestionScoreRespDTO.class)))  }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("candidate/{candidateId}/assessment/{assessmentId}/score")
    public List<AssessmentQuestionScoreRespDTO> getCandidateScoresByAssessment(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @PathVariable("assessmentId") @Positive Integer assessmentId) {
        return this.gradingService.getCandidateScoresByAssessment(candidateId, assessmentId);
    }

    @Operation(summary = "Get candidate question scores for assessment block")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", array = @ArraySchema(schema = @Schema(implementation = AssessmentQuestionScoreRespDTO.class)))  }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_ASSESSMENT_BLOCK_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("candidate/{candidateId}/assessment-block/{assessmentBlockId}/score")
    public List<AssessmentQuestionScoreRespDTO> getCandidateScoresByAssessmentBlock(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @PathVariable("assessmentBlockId") @Positive Integer assessmentBlockId){
        return this.gradingService.getCandidateScoresByAssessmentBlock(candidateId, assessmentBlockId);
    }

    @Operation(summary = "Get candidate unscored questions for assessment")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", array = @ArraySchema(schema = @Schema(implementation = Integer.class))) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("candidate/{candidateId}/assessment/{assessmentId}/question/unscored")
    public List<Integer> getCandidateUnScoredQuestionsForAssessment(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @PathVariable("assessmentId") @Positive Integer assessmentId) {
        return  this.gradingService.getCandidateUnScoredQuestionsForAssessment(candidateId, assessmentId);
    }

    @Operation(summary = "Get candidate question comments")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", array = @ArraySchema(schema = @Schema(implementation = QuestionCommentRespDTO.class))) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_QUESTION_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("candidate/{candidateId}/question/{questionId}/comment")
    public List<QuestionCommentRespDTO> getCandidateQuestionComments(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @PathVariable("questionId") @Positive Integer questionId
            ) {
        return  this.gradingService.getCandidateQuestionComments(candidateId, questionId);
    }

    @Operation(summary = "Submit candidate question comments")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_QUESTION_ID_DESCRIPTION, content = @Content)
    })
    @PutMapping("candidate/{candidateId}/question/{questionId}/comment")
    public void saveCandidateQuestionComment(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @PathVariable("questionId") @Positive Integer questionId,
            @RequestBody() @Valid QuestionCommentReqDTO questionCommentReqDto) {
       this.gradingService.saveCandidateQuestionComment(candidateId, questionId, questionCommentReqDto);
    }

    @Operation(summary = "Submit candidate question score")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_QUESTION_ID_DESCRIPTION, content = @Content)
    })
    @PutMapping("candidate/{candidateId}/score")
    public void saveCandidateResponseScore(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @RequestBody() @Valid QuestionScoreReqDTO questionScoreReqDto) {
        this.gradingService.saveCandidateResponseScore(candidateId, questionScoreReqDto);
    }

    @Operation(summary = "Mark assessment grading complete")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @PatchMapping("candidate/{candidateId}/assessment/{assessmentId}/grading/complete")
    public void markCandidateAssessmentGradingComplete(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @PathVariable("assessmentId") @Positive Integer assessmentId
    ) {
        this.gradingService.markCandidateAssessmentGradingComplete(candidateId, assessmentId);
    }

    @Operation(summary = "Get candidate assessment decision")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = AssessmentDecision.class))} ),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("candidate/{candidateId}/assessment/{assessmentId}/decision")
    public AssessmentDecision getCandidateAssessmentDecision(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @PathVariable("assessmentId") @Positive Integer assessmentId){
        return this.gradingService.getCandidateAssessmentDecision(candidateId, assessmentId);
    }

    @Operation(summary = "Set candidate assessment decision")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = GradingResponseCodes.INVALID_ASSESSMENT_DECISION_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = GradingResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @PutMapping("candidate/{candidateId}/assessment/{assessmentId}/decision/{decision}")
    public void setCandidateAssessmentDecision(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @PathVariable("assessmentId") @Positive Integer assessmentId,
            @PathVariable("decision") @NotEmpty String decision) {
        this.gradingService.setCandidateAssessmentDecision(candidateId, assessmentId, decision);
    }

}

File: javascript/smartlist/smartlist-api-mysql-project/src/modules/grading/grading.controller.ts
import {
  Body,
  Controller,
  Get,
  HttpCode,
  Param,
  ParseIntPipe,
  Patch,
  Put,
} from '@nestjs/common';
import { ApiOperation, ApiResponse } from '@nestjs/swagger';
import { AssessmentDecision } from 'src/enums/assessment.decision';
import { Role } from 'src/enums/role';
import { Roles } from 'src/guards/roles.decorator';
import { QuestionCommentReqDto } from './dto/request/question-comment.req.dto';
import { QuestionScoreReqDto } from './dto/request/question-score.req.dto';
import { AssessmentQuestionScoreRespDto } from './dto/response/assessment-question-score.resp.dto';
import { QuestionCommentRespDto } from './dto/response/question-comment.resp.dto';
import { GradingResponseCodes } from './grading.response.codes';
import { GradingService } from './grading.service';

@Controller('grading')
@Roles(Role.SUPERADMIN, Role.ADMIN, Role.MEMBER)
export class GradingController {
  constructor(private readonly gradingService: GradingService) {}

  @ApiOperation({ summary: 'Get candidate question scores for assessment' })
  @ApiResponse({
    ...GradingResponseCodes.SUCCESS,
    type: [],
  })
  @ApiResponse(GradingResponseCodes.BAD_REQUEST)
  @ApiResponse(GradingResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(GradingResponseCodes.INVALID_ASSESSMENT_ID)
  @HttpCode(200)
  @Get('candidate/:candidateId/assessment/:assessmentId/score')
  async getScoresForAssessment(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Param('assessmentId', ParseIntPipe) assessmentId: number,
  ): Promise<AssessmentQuestionScoreRespDto[]> {
    return await this.gradingService.getCandidateScoresByAssessment(
      candidateId,
      assessmentId,
    );
  }

  @ApiOperation({
    summary: 'Get candidate question scores for assessment block',
  })
  @ApiResponse({
    ...GradingResponseCodes.SUCCESS,
    type: [],
  })
  @ApiResponse(GradingResponseCodes.BAD_REQUEST)
  @ApiResponse(GradingResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(GradingResponseCodes.INVALID_ASSESSMENT_BLOCK_ID)
  @HttpCode(200)
  @Get('candidate/:candidateId/assessment-block/:assessmentBlockId/score')
  async getScoresForAssessmentBlock(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Param('assessmentBlockId', ParseIntPipe) assessmentBlockId: number,
  ): Promise<AssessmentQuestionScoreRespDto[]> {
    return await this.gradingService.getCandidateScoresByAssessmentBlock(
      candidateId,
      assessmentBlockId,
    );
  }

  @ApiOperation({
    summary: 'Get candidate unscored questions for assesment',
  })
  @ApiResponse({
    ...GradingResponseCodes.SUCCESS,
    type: [],
  })
  @ApiResponse(GradingResponseCodes.BAD_REQUEST)
  @ApiResponse(GradingResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(GradingResponseCodes.INVALID_ASSESSMENT_ID)
  @HttpCode(200)
  @Get('candidate/:candidateId/assessment/:assessmentId/question/unscored')
  async getUnScoredQuestionsForAssessment(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Param('assessmentId', ParseIntPipe) assessmentId: number,
  ): Promise<number[]> {
    return await this.gradingService.getCandidateUnScoredQuestionsForAssessment(
      candidateId,
      assessmentId,
    );
  }

  @ApiOperation({
    summary: 'Get candidate question comments',
  })
  @ApiResponse({
    ...GradingResponseCodes.SUCCESS,
    type: [],
  })
  @ApiResponse(GradingResponseCodes.BAD_REQUEST)
  @ApiResponse(GradingResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(GradingResponseCodes.INVALID_QUESTION_ID)
  @HttpCode(200)
  @Get('candidate/:candidateId/question/:questionId/comment')
  async getAllQuestionComments(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Param('questionId', ParseIntPipe) questionId: number,
  ): Promise<QuestionCommentRespDto[]> {
    return await this.gradingService.getCandidateQuestionComments(
      candidateId,
      questionId,
    );
  }

  @ApiOperation({ summary: 'Submit candidate question comments' })
  @ApiResponse({
    ...GradingResponseCodes.SUCCESS,
    type: [],
  })
  @ApiResponse(GradingResponseCodes.BAD_REQUEST)
  @ApiResponse(GradingResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(GradingResponseCodes.INVALID_QUESTION_ID)
  @HttpCode(200)
  @Put('candidate/:candidateId/question/:questionId/comment')
  async setQuestionComment(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Param('questionId', ParseIntPipe) questionId: number,
    @Body()
    questionCommentReqDto: QuestionCommentReqDto,
  ) {
    return await this.gradingService.saveCandidateQuestionComment(
      candidateId,
      questionId,
      questionCommentReqDto,
    );
  }

  @ApiOperation({ summary: 'Submit candidate question score' })
  @ApiResponse({
    ...GradingResponseCodes.SUCCESS,
    type: [],
  })
  @ApiResponse(GradingResponseCodes.BAD_REQUEST)
  @ApiResponse(GradingResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(GradingResponseCodes.INVALID_ASSESSMENT_ID)
  @ApiResponse(GradingResponseCodes.INVALID_QUESTION_ID)
  @HttpCode(200)
  @Put('candidate/:candidateId/score')
  async saveCandidateResponseScore(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Body()
    questionScoreReqDto: QuestionScoreReqDto,
  ) {
    return await this.gradingService.saveCandidateResponseScore(
      candidateId,
      questionScoreReqDto,
    );
  }

  @ApiOperation({
    summary: 'Get candidate assessment decision',
  })
  @ApiResponse({
    ...GradingResponseCodes.SUCCESS,
    type: [],
  })
  @ApiResponse(GradingResponseCodes.BAD_REQUEST)
  @ApiResponse(GradingResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(GradingResponseCodes.INVALID_ASSESSMENT_BLOCK_ID)
  @HttpCode(200)
  @Get('candidate/:candidateId/assessment/:assessmentId/decision')
  async getAssessmentDecision(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Param('assessmentId', ParseIntPipe) assessmentId: number,
  ): Promise<AssessmentDecision> {
    return await this.gradingService.getCandidateAssessmentDecision(
      candidateId,
      assessmentId,
    );
  }

  @ApiOperation({ summary: 'Mark assessment grading complete' })
  @ApiResponse(GradingResponseCodes.SUCCESS)
  @ApiResponse(GradingResponseCodes.BAD_REQUEST)
  @ApiResponse(GradingResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(GradingResponseCodes.INVALID_ASSESSMENT_ID)
  @HttpCode(200)
  @Patch('candidate/:candidateId/assessment/:assessmentId/grading/complete')
  async markAssessmentGradingComplete(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Param('assessmentId', ParseIntPipe) assessmentId: number,
  ) {
    return await this.gradingService.markCandidateAssessmentGradingComplete(
      candidateId,
      assessmentId,
    );
  }

  @ApiOperation({ summary: 'Set candidate assessment decision' })
  @ApiResponse({
    ...GradingResponseCodes.SUCCESS,
    type: [],
  })
  @ApiResponse(GradingResponseCodes.BAD_REQUEST)
  @ApiResponse(GradingResponseCodes.INVALID_ASSESSMENT_DECISION)
  @ApiResponse(GradingResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(GradingResponseCodes.INVALID_ASSESSMENT_ID)
  @HttpCode(200)
  @Put('candidate/:candidateId/assessment/:assessmentId/decision/:decision')
  async setAssessmentDecision(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Param('assessmentId', ParseIntPipe) assessmentId: number,
    @Param('decision') decision: string,
  ) {
    return await this.gradingService.setCandidateAssessmentDecision(
      candidateId,
      assessmentId,
      decision,
    );
  }
}
